### PR TITLE
[SPARK-32559][SQL]Fix the trim logic in UTF8String.toInt/toLong did't handle non-ASCII characters correctly

### PR DESCRIPTION
**Pull request by WangGuangxin**
### What changes were proposed in this pull request?
The trim logic in Cast expression introduced in https://github.com/apache/spark/pull/26622 trim non-ASCII characters unexpectly.

Before this patch
image

After this patch
image

### Why are the changes needed?
The behavior described above doesn't make sense, and also doesn't consistent with the behavior when cast a string to double/float, as well as doesn't consistent with the behavior of Hive

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Added more UT

### CODE REVIEW

**Comment by WangGuangxin** [comment]
@cloud-fan @yaooqinn @gengliangwang Could you please help review this?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -575,14 +575,14 @@ public UTF8String trim() {
   public UTF8String trimAll() {
     int s = 0;
     // skip all of the whitespaces (<=0x20) in the left side
-    while (s < this.numBytes && getByte(s) <= ' ') s++;
+    while (s < this.numBytes && canBeTrimmed(getByte(s))) s++;
```
Comment (yaooqinn) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:578]: Does`java.lang.Character.isWhitespace(char)` work as well?

Comment (WangGuangxin) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:578]: yes, that's a good choice

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
@WangGuangxin  can you open a backport PR for 3.0?

**Comment by WangGuangxin** [comment]
> @WangGuangxin can you open a backport PR for 3.0?

sure

File: sql/core/src/test/resources/sql-tests/inputs/cast.sql [hunk 2]
```diff
@@ -70,6 +70,11 @@ select cast(' 1' as bigint);
 select cast(' 1' as float);
 select cast(' 1 ' as DOUBLE);
 select cast('1.0 ' as DEC);
+select cast('1中文' as tinyint);
```
Comment (srowen) [inline sql/core/src/test/resources/sql-tests/inputs/cast.sql:73]: This is purely to educate me, but those characters are considered whitespace?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/cast.sql:73]: This kind of character needs multiple bytes, so `getByte(s) <= ' '` may not work.

**Pull request closed without merging**